service/history/replication/dynamic_task_batch_sizer.go

```diff
@@ -0,0 +1,105 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package replication
+
+import (
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/rangeiter"
+	"github.com/uber/cadence/service/history/config"
+)
+
+// DynamicTaskBatchSizer is responsible for the batch size used to retrieve ReplicationTasks by TaskAckManager
+// It adjusts the task batch size based on the error and the getTasksResult,
+// and use the following rules:
+//
+//  1. If there is an error, decrease the task batch size.
+//     In case of an increased load to the database, we should reduce the load to the database
+//
+//  2. If the task batch size is shrunk, decrease the task batch size.
+//     The payload size of messages is too big, so we should decrease
+//     the number of messages to be sure that future messages will not be shrunk
+//
+//  3. If the read level of a passive cluster has not been changed and there are no fetched tasks,
+//     not changed the task batch size. There is no need to change because the replication is not stuck,
+//     there just are no new tasks
+//
+//  4. If the read level of a passive cluster has not been changed and if there are fetched tasks,
+//     decrease the task batch size. There were no new replication tasks in the previous call
+//     or the replication was stuck on the passive side
+//
+//  5. If the read level of a passive cluster has been changed and if there are more tasks in db,
+//     increase the task batch size. We should retrieve the maximum possible value at the next time,
+//     as there are more tasks to be replicated
+//
+//  6. If the read level of a passive cluster has been changed and if there are no more tasks in db,
+//     not change the size. The existing size is already enough, and there are no more tasks to be replicated
+type DynamicTaskBatchSizer interface {
+	analyse(err error, state *getTasksResult)
+	value() int
+}
+
+// dynamicTaskBatchSizerImpl is the implementation of DynamicTaskBatchSizer
+type dynamicTaskBatchSizerImpl struct {
+	iter   rangeiter.Iterator[int]
+	logger log.Logger
+}
+
+// NewDynamicTaskBatchSizer creates a new dynamicTaskBatchSizerImpl
+func NewDynamicTaskBatchSizer(shardID int, logger log.Logger, config *config.Config) DynamicTaskBatchSizer {
+	logger = logger.WithTags(tag.ComponentReplicationDynamicTaskBatchSizer)
+	return &dynamicTaskBatchSizerImpl{
+		logger: logger,
+		iter: rangeiter.NewDynamicConfigLinearIterator(
+			func() int { return config.ReplicatorProcessorMinTaskBatchSize(shardID) },
+			func() int { return config.ReplicatorProcessorMaxTaskBatchSize(shardID) },
+			func() int { return config.ReplicatorProcessorBatchSizeStepCount(shardID) },
+			logger,
+		),
+	}
+}
+
+func (d *dynamicTaskBatchSizerImpl) analyse(err error, state *getTasksResult) {
+	switch {
+	case err != nil:
+		d.logger.Debug(
+			"Decrease task batch size due to error",
```

Comment:
it'd be great to get some counters on this so that we can analyize if we see problems after the fact (debug logs having to be turned on and all)

Reply:
Introduced new metrics